File: examples/governance/compliance_domain/lib/compliance_domain/regulatory_framework/commands/register_framework.rb
module ComplianceDomain
  class RegulatoryFramework
    module Commands
      class RegisterFramework
        include Hecks::Command
        emits "RegisteredFramework"

        attr_reader :name
        attr_reader :jurisdiction
        attr_reader :version
        attr_reader :authority

        def initialize(
          name: nil,
          jurisdiction: nil,
          version: nil,
          authority: nil
        )
          @name = name
          @jurisdiction = jurisdiction
          @version = version
          @authority = authority
        end

        def call
          RegulatoryFramework.new(
            name: name,
            jurisdiction: jurisdiction,
            version: version,
            authority: authority,
            status: "draft"
          )
        end
      end
    end
  end
end

File: examples/governance_static_go/domain/register_framework.go
package domain

import (
	"time"
)

type RegisterFramework struct {
	Name string `json:"name"`
	Jurisdiction string `json:"jurisdiction"`
	Version string `json:"version"`
	Authority string `json:"authority"`
}

func (c RegisterFramework) CommandName() string { return "RegisterFramework" }

func (c RegisterFramework) Execute(repo RegulatoryFrameworkRepository) (*RegulatoryFramework, *RegisteredFramework, error) {
	agg := NewRegulatoryFramework(c.Name, c.Jurisdiction, c.Version, time.Time{}, c.Authority, nil, "")
	if err := agg.Validate(); err != nil {
		return nil, nil, err
	}
	if err := repo.Save(agg); err != nil {
		return nil, nil, err
	}
	event := RegisteredFramework{
		AggregateID: agg.ID,
		Name: c.Name,
		Jurisdiction: c.Jurisdiction,
		Version: c.Version,
		Authority: c.Authority,
		OccurredAt: time.Now(),
	}
	return agg, &event, nil
}
